File: common/src/main/java/com/ssblur/scriptor/helpers/PlayerSpellsSavedData.java
package com.ssblur.scriptor.helpers;

import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.client.renderer.texture.SpriteContents;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtOps;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.saveddata.SavedData;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;

public class PlayerSpellsSavedData extends SavedData {
  public HashMap<Integer, HashMap<String, Boolean>> spells = new HashMap<>();
  public static final Codec<PlayerSpellsSavedData> CODEC = RecordCodecBuilder.create(instance -> instance.group(
    Codec.STRING.listOf()
      .fieldOf("spells")
      .forGetter(
        data -> {
          ArrayList<String> list = new ArrayList<>();
          for(int i: data.spells.keySet())
            for(var s: data.spells.get(i).keySet())
              list.add(i + ":" + s);
          return list;
        }
      )
  ).apply(instance, PlayerSpellsSavedData::new));

  public PlayerSpellsSavedData(List<String> pairs) {
    for(var pair: pairs) {
      var split = pair.split(":", 2);
      int first = Integer.parseInt(split[0]);
      if(!spells.containsKey(first))
        spells.put(first, new HashMap<>());
      spells.get(first).put(split[1], true);
    }
  }

  public PlayerSpellsSavedData() {}

  public HashMap<String, Boolean> getTier(int tier) {
    if(!spells.containsKey(tier))
      spells.put(tier, new HashMap<>());
    return spells.get(tier);
  }

  @Override
  public CompoundTag save(CompoundTag tag) {
    var result = CODEC.encodeStart(NbtOps.INSTANCE, this).get().left();
    result.ifPresent(value -> tag.put("scriptor:obtained_spells", value));
    return tag;
  }

  public static PlayerSpellsSavedData load(CompoundTag tag) {
    var input = tag.get("scriptor:obtained_spells");
    if(input != null) {
      var result = CODEC.decode(NbtOps.INSTANCE, input).get().left();
      if(result.isPresent() && result.get().getFirst() != null)
        return result.get().getFirst();
    }
    return null;
  }

  /**
   * A helper for calling computeIfAbsent for this class from the Overworld
   * @param player Any ServerPlayer
   * @return The DictionarySavedData for this world
   */
  @Nullable
  public static PlayerSpellsSavedData computeIfAbsent(Player player) {
    MinecraftServer minecraft;
    var level = player.level();
    minecraft = level.getServer();
    if(minecraft == null) return null;

    var server = minecraft.getLevel(Level.OVERWORLD);
    if(server == null) return null;

    try {
      var storage = server.getDataStorage();
      if(storage == null) return null;

      if(!Files.exists(storage.dataFolder.toPath().resolve("scriptor_players")))
        Files.createDirectory(storage.dataFolder.toPath().resolve("scriptor_players"));
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    Objects.requireNonNull(server);
    return server.getDataStorage().computeIfAbsent(
      PlayerSpellsSavedData::load,
      PlayerSpellsSavedData::new,
      "scriptor_players/spells_" + player.getStringUUID()
    );
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/data/saved_data/PlayerSpellsSavedData.kt
package com.ssblur.scriptor.data.saved_data

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import com.ssblur.scriptor.mixin.DimensionDataStorageAccessor
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtOps
import net.minecraft.util.datafix.DataFixTypes
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.saveddata.SavedData
import java.io.IOException
import java.nio.file.Files
import java.util.*

class PlayerSpellsSavedData: SavedData {
  var spells: HashMap<Int, HashMap<String, Boolean>> = HashMap()
  var scraps: HashMap<Int, HashMap<String, Boolean>> = HashMap()

  constructor(spellList: List<String>, scrapList: List<String>) {
    for (pair in spellList) {
      val split = pair.split(":".toRegex(), limit = 2).toTypedArray()
      val first = split[0].toInt()
      if (!spells.containsKey(first)) spells[first] = HashMap()
      spells[first]!![split[1]] = true
    }

    for (pair in scrapList) {
      val split = pair.split(":".toRegex(), limit = 2).toTypedArray()
      val first = split[0].toInt()
      if (!spells.containsKey(first)) spells[first] = HashMap()
      spells[first]!![split[1]] = true
    }
  }

  constructor()

  fun getTier(tier: Int): HashMap<String, Boolean> {
    if (!spells.containsKey(tier)) spells[tier] = HashMap()
    return spells[tier]!!
  }

  fun getScrapTier(tier: Int): HashMap<String, Boolean> {
    if (!scraps.containsKey(tier)) scraps[tier] = HashMap()
    return scraps[tier]!!
  }

  override fun save(tag: CompoundTag, provider: HolderLookup.Provider): CompoundTag {
    CODEC.encodeStart(NbtOps.INSTANCE, this).ifSuccess { tag.put("scriptor:obtained_spells", it) }
    return tag
  }

  companion object {
    val CODEC: Codec<PlayerSpellsSavedData?> =
      RecordCodecBuilder.create { instance: RecordCodecBuilder.Instance<PlayerSpellsSavedData?> ->
        instance.group(
          Codec.STRING.listOf()
            .fieldOf("spells")
            .forGetter { data: PlayerSpellsSavedData? ->
              val list = ArrayList<String>()
              for (i in data!!.spells.keys) for (s in data.spells[i]!!.keys) list.add("$i:$s")
              list
            },
          Codec.STRING.listOf()
            .fieldOf("scraps")
            .forGetter {
              it?.let {
                val list = ArrayList<String>()
                for (i in it.scraps.keys) for (s in it.scraps[i]!!.keys) list.add("$i:$s")
                list
              }
            }
        ).apply(instance) { spellList: List<String>, scrapList: List<String> ->
          PlayerSpellsSavedData(
            spellList,
            scrapList
          )
        }
      }

    fun load(tag: CompoundTag, @Suppress("unused_parameter") provider: HolderLookup.Provider?): PlayerSpellsSavedData? {
      val input = tag["scriptor:obtained_spells"]
      if (input != null) {
        val result = CODEC.decode(NbtOps.INSTANCE, input).result()
        if (result.isPresent && result.get().first != null) return result.get().first
      }
      return null
    }

    /**
     * A helper for calling computeIfAbsent for this class from the Overworld
     * @param player Any ServerPlayer
     * @return The DictionarySavedData for this world
     */
    @JvmStatic
    fun computeIfAbsent(player: Player): PlayerSpellsSavedData? {
      val level = player.level()
      val minecraft = level.server ?: return null
      val server = minecraft.getLevel(Level.OVERWORLD) ?: return null

      try {
        val storage = server.dataStorage as DimensionDataStorageAccessor

        if (!Files.exists(storage.dataFolder.toPath().resolve("scriptor_players"))) Files.createDirectory(
          storage.dataFolder.toPath().resolve("scriptor_players")
        )
      } catch (e: IOException) {
        throw RuntimeException(e)
      }
      Objects.requireNonNull(server)
      return server.dataStorage.computeIfAbsent(
        Factory(
          { PlayerSpellsSavedData() },
          { tag: CompoundTag, provider: HolderLookup.Provider? -> load(tag, provider) },
          DataFixTypes.SAVED_DATA_MAP_DATA
        ),
        "scriptor_players/spells_" + player.stringUUID
      )
    }
  }
}
